tests/Test.ADAL.NET.Unit/ClientAssertionCertificateTests.cs

```diff
@@ -0,0 +1,176 @@
+using Test.ADAL.Common;
+using Test.ADAL.NET.Common;
+using Test.ADAL.NET.Common.Mocks;
+
+namespace Test.ADAL.NET.Unit
+{
+    [TestClass]
+    [DeploymentItem("valid_cert.pfx")]
+    public class ClientAssertionCertificateTests
+    {
+        private PlatformParameters platformParameters;
+
+        [TestInitialize]
+        public void Initialize()
+        {
+            HttpMessageHandlerFactory.ClearMockHandlers();
+            InstanceDiscovery.InstanceCache.Clear();
+            HttpMessageHandlerFactory.AddMockHandler(MockHelpers.CreateInstanceDiscoveryMockHandler(TestConstants.GetDiscoveryEndpoint(TestConstants.DefaultAuthorityCommonTenant)));
+            platformParameters = new PlatformParameters(PromptBehavior.Auto);
+        }
+
+        [TestMethod]
+        [Description("Test for Client assertion with X509")]
+        public async Task ClientAssertionWithX509Test()
+        {
+            var certificate = new X509Certificate2("valid_cert.pfx", TestConstants.DefaultPassword);
+            var clientAssertion = new ClientAssertionCertificate(TestConstants.DefaultClientId, certificate);
+
+            var context = new AuthenticationContext(TestConstants.DefaultAuthorityCommonTenant, new TokenCache());
+            var expectedAudience = TestConstants.DefaultAuthorityCommonTenant + "oauth2/token";
+
+            var validCertClaim = "\"x5c\":\"" + Base64UrlEncoder.Encode(certificate.ToString());
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler(TestConstants.GetTokenEndpoint(TestConstants.DefaultAuthorityCommonTenant))
+            {
+                Method = HttpMethod.Post,
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent("{\"token_type\":\"Bearer\",\"expires_in\":\"3599\",\"access_token\":\"some-access-token\"}")
+                },
+                PostData = new Dictionary<string, string>
+                {
+                    {"client_id", TestConstants.DefaultClientId},
+                    {"grant_type", "client_credentials"},
+                    {"client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"}
+                },
+                AdditionalRequestValidation = request =>
+                {
+                    var requestContent = request.Content.ReadAsStringAsync().GetAwaiter().GetResult();
+                    var formsData = EncodingHelper.ParseKeyValueList(requestContent, '&', true, null);
+
+                    // Check presence of client_assertion in request
+                    string encodedJwt;
+                    Assert.IsTrue(formsData.TryGetValue("client_assertion", out encodedJwt), "Missing client_assertion from request");
+
+                    // Check presence of x5c cert claim
+                    var jwt = EncodingHelper.Base64Decode(encodedJwt.Split('.')[0]);
+                    Assert.IsTrue(jwt.Contains(validCertClaim));
```

Comment:
you should add a unit test for JsonWebToken class (in unit test project) to see that the certificate is correctly put in as a Base64 encoded value.